docs/extensions/shoot-health-status-conditions.md

```diff
@@ -0,0 +1,84 @@
+
+* `APIServerAvailable`: This type indicates whether the shoot's kube-apiserver is available or not.
+* `ControlPlaneHealthy`: This type indicates whether all the control plane components deployed to the shoot's namespace in the seed do exist and are running fine.
+* `EveryNodeReady`: This type indicates whether all `Node`s and all `Machine` objects report healthiness.
+* `SystemComponentsHealthy`: This type indicates whether all system components deployed to the `kube-system` namespace in the shoot do exist and are running fine.
+
+Every `Shoot` resource has a `status.conditions[]` list that contain the mentioned types, together with a `status` (`True`/`False`) and a descriptive message/explanation of the `status`.
+
+Most extension controllers are deploying components and resources as part of their reconciliation flows into the seed or shoot cluster.
+A prominent example for this is the `ControlPlane` controller that usually deploys a cloud-controller-manager or CSI controllers as part of the shoot control plane.
+Now that the extensions deploy resources into the cluster, especially resources that are essential for the functionality of the cluster, they might want to contribute to Gardener's checks mentioned above.
+
+## What can extensions do to contribute to Gardener's health checks?
+
+Every extension resource in Gardener's `extensions.gardener.cloud/v1alpha1` API group also has a `status.conditions[]` list (like the `Shoot`).
+Extension controllers can write conditions to the resource they are acting on and use a type that also exist in the shoot's conditions.
+One exception is that `APIServerAvailable` can't be used as the Gardener clearly can identify the status of this condition and it doesn't make sense for extension to try to contribute/modify it.
+
+As an example for the `ControlPlane` controller let's take a look at following resource:
+
+```yaml
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: ControlPlane
+metadata:
+  name: control-plane
+  namespace: shoot--foo--bar
+spec:
+  ...
+status:
+  conditions:
+  - type: ControlPlaneHealthy
+    status: "False"
+    reason: DeploymentUnhealthy
+    message: 'Deployment cloud-controller-manager is unhealthy: condition "Available" has
+      invalid status False (expected True) due to MinimumReplicasUnavailable: Deployment
+      does not have minimum availability.'
+    lastUpdateTime: "2014-05-25T12:44:27Z"
+  - type: ConfigComputedSuccessfully
+    status: "True"
+    reason: ConfigCreated
+    message: The cloud-provider-config has been successfully computed.
+    lastUpdateTime: "2014-05-25T12:43:27Z"
+```
+
+The extension controller has declared in its extension resource that one if the deployments it is responsible for is not healthy.
+Also, it has written a second condition using a type that is not known by Gardener.
+
+Gardener will pick the list of conditions up and see that the there is one with a type `ControlPlaneHealthy`.
```

Comment:
Should we also consider other conditions in the extensions, for example in this case: `ConfigComputedSuccessfully`, I though that if an exetnsion is categorized as a controlplane extensions then all its conditions should contribute and reflect in the `ControlPlaneHealthy` condition or do you aassume it is the task of the extension controller to aggregate all the sub-conditions and report at the end only one of the four main conditions?

Reply:
The `ControlPlane` extension may also deploy system components into the shoot, e.g. for Alicloud the CSI driver plugin is deployed. Consequently, it might also want to contribute to the `SystemComponentsHealthy` condition.
Bottom line: We can't make such an assumption.

> to aggregate all the sub-conditions and report at the end only one of the four main conditions?

I think yes, that's what would be needed.